File: Python/Trapping rain water.py
def TrappingRainWater(heights):
    if not heights:
        return 0

    l, r = 0, len(heights)-1
    left_max, right_max = heights[l], heights[r]

    ans = 0

    while l < r:
        if left_max < right_max:
            l += 1
            left_max = max(left_max, heights[l])
            ans += left_max - heights[l]
        else:
            r -= 1
            right_max = max(right_max, heights[r])
            ans += right_max - heights[r]
    return ans


print(TrappingRainWater([0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1]))

File: Java/trappingRainWater.java
package Java;

/**
 * trappingRainWater
 */
public class trappingRainWater {
    public static int Trapping_Rain_Water(int[] heights) {
        if (heights == null || heights.length == 0) {
            return 0;
        }
        int l = 0;
        int r = heights.length - 1;
        int leftMax = heights[l];
        int rightMax = heights[r];

        int res = 0;

        while (l < r) {
            if (leftMax < rightMax) {
                l += 1;
                leftMax = Math.max(leftMax, heights[l]);
                res += leftMax - heights[l];
            } else {
                r -= 1;
                rightMax = Math.max(rightMax, heights[r]);
                res += rightMax - heights[r];
            }
        }
        return res;

    }

    public static void main(String[] args) {
        int[] heights = { 0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1 };
        System.out.println(Trapping_Rain_Water(heights));
    }
}